Add ``--output`` parameter to demo to save results

GalyaZalesskaya opened the pull request:
<!-- Contributing guide: https://github.com/openvinotoolkit/training_extensions/blob/develop/CONTRIBUTING.md -->

### Summary

- Added the ability to save inference images in OTX CLI demo, exportable demo with ``--save-results-to`` parameter. 
It helps to keep and check the inference results after running a demo, using the command line on a remote station, without the necessity to use X server. 
- Checked for images, videos, image folders for Classification task

Jira ticket: https://jira.devtools.intel.com/browse/CVS-107856

### How to test



### Checklist

<!-- Put an 'x' in all the boxes that apply -->

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [x] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [x] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [x] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

Commit:
Add --save-results-to for demo CLI and exportable

Commit:
Fix linter issues

Commit:
Update docs & CHANGELOG

Commit:
Fix linter

Commit:
Update exportable code requirements otx commit

GalyaZalesskaya commented on otx/api/usecases/exportable_code/demo/requirements.txt:
```diff
@@ -1,4 +1,4 @@
 openvino==2022.3.0
 openmodelzoo-modelapi==2022.3.0
-otx @ git+https://github.com/openvinotoolkit/training_extensions/@dd03235da2319815227f1b75bce298ee6e8b0f31#egg=otx
+otx @ git+https://github.com/openvinotoolkit/training_extensions/@8c11c3d42c726e6e0eda7364f00cf8ed4dbdc2e9#egg=otx
```
@goodsong81 The previous and updated commit is not a part of OTX develop, but from the fork. Can you please clarify if we need to update this commit again, then these changes will be merged.

goodsong81 replied:
As far as pip install works, it's OK as we'll anyway update this commit to a PyPI version at the time of release.
![image](https://user-images.githubusercontent.com/4629606/231624631-3a12eab2-4051-49f7-8744-4da7d1f3b4a7.png)

GalyaZalesskaya commented:
Actually, do I understand correctly, that this parameter doesn't require adding extra tests?

eunwoosh commented on docs/source/guide/get_started/quick_start_guide/cli_commands.rst:
```diff
@@ -474,10 +474,10 @@ Demonstration
 
     positional arguments:
-      template              Enter the path or ID or name of the template file.
+      template              Enter the path or ID or name of the template file. 
```
```suggestion
      template              Enter the path or ID or name of the template file.
```

eunwoosh commented on otx/api/usecases/exportable_code/demo/demo_package/utils.py:
```diff
@@ -47,9 +47,9 @@ def create_output_converter(task_type: TaskType, labels: LabelSchemaEntity):
 
 
-def create_visualizer(_task_type: TaskType, no_show: bool = False):
+def create_visualizer(_task_type: TaskType, no_show: bool = False, save_results_to: str = None):
```
```suggestion
def create_visualizer(_task_type: TaskType, no_show: bool = False, save_results_to: Optional[str] = None):
```

eunwoosh commented on otx/api/usecases/exportable_code/visualizers/visualizer.py:
```diff
@@ -59,13 +60,15 @@ class Visualizer(IVisualizer):
         is_one_label: bool = False,
         no_show: bool = False,
         delay: Optional[int] = None,
+        save_results_to: str = None,
```
```suggestion
        save_results_to: Optional[str] = None,
```

eunwoosh commented on otx/api/usecases/exportable_code/visualizers/visualizer.py:
```diff
@@ -59,13 +60,15 @@ class Visualizer(IVisualizer):
         no_show: bool = False,
         delay: Optional[int] = None,
+        save_results_to: str = None,
+        saved_frames: Dict[str, list] = None,
```
```suggestion
        saved_frames: Optional[Dict[str, list]] = None,
```

eunwoosh commented on otx/api/usecases/exportable_code/visualizers/visualizer.py:
```diff
@@ -111,3 +119,48 @@ def is_quit(self) -> bool:
 
         return ord("q") == cv2.waitKey(self.delay)
+
+    def save_frame(self, image, input_path, streamer_type) -> None:
```
Could you add type hint?

GalyaZalesskaya replied:
Added type hints to all functions I've added

eunwoosh commented on otx/api/usecases/exportable_code/visualizers/visualizer.py:
```diff
@@ -111,3 +119,48 @@ def is_quit(self) -> bool:
+            else:
+                self.saved_frames[filename] = image
+
+    def dump_frames(self, streamer) -> None:
```
Could you add type hint?

eunwoosh commented on otx/api/usecases/exportable_code/visualizers/visualizer.py:
```diff
@@ -111,3 +119,48 @@ def is_quit(self) -> bool:
+        Args:
+            streamer (str): The streamer with images to be saved
+        """
+        if len(self.saved_frames) > 0 and self.save_results_to:
```
```suggestion
        if self.saved_frames and self.save_results_to:
```
Although It doesn't matter, this way is recommend to check whether dict is empty :)

GalyaZalesskaya replied:
Thank you :)

eunwoosh commented on otx/api/usecases/exportable_code/visualizers/visualizer.py:
```diff
@@ -111,3 +119,48 @@ def is_quit(self) -> bool:
+        """Save frames to file system.
+
+        Args:
+            streamer (str): The streamer with images to be saved
```
Is it right that streamer is str type?

GalyaZalesskaya replied:
Removed

eunwoosh commented on otx/api/usecases/exportable_code/visualizers/visualizer.py:
```diff
@@ -111,3 +119,48 @@ def is_quit(self) -> bool:
+                for frame in frames:
+                    out.write(frame)
+                out.release()
+                print(f"Video was saved to {video_path}")
```
Isn't it better to use logger.info?

GalyaZalesskaya replied:
Done

eunwoosh commented on otx/api/usecases/exportable_code/visualizers/visualizer.py:
```diff
@@ -111,3 +119,48 @@ def is_quit(self) -> bool:
+                for filename, frame in self.saved_frames.items():
+                    image_path = str(output_path / filename)
+                    cv2.imwrite(image_path, frame)
+                    print(f"Image was saved to {image_path}")
```
same as above

eunwoosh commented on otx/api/usecases/exportable_code/demo/demo_package/executors/asynchronous.py:
```diff
@@ -39,12 +39,13 @@ def run(self, input_stream: Union[int, str], loop: bool = False) -> None:
         stop_visualization = False
 
-        for frame in streamer:
+        for (frame, input_path) in streamer:
```
I think it can raise an error if streamer is `ThreadStreamer`.

sovrasov replied:
We can stick to a pre-defined output video/images names to avoid introducing the additional output from streamer.

GalyaZalesskaya replied:
Sure, it will be much easier. But it would cause a problem, that input and output names aren't consistent and it's hard to identify, which image was the source, based on naming

But if additional output in streamer is highly unwelcome, I can write a standalone function to get input names from args.input, without changing the streamer code.

eunwoosh reviewed (commented):
Thanks for your work! I left some comment. please take a look

harimkang commented on docs/source/guide/get_started/quick_start_guide/cli_commands.rst:
```diff
@@ -474,10 +474,10 @@ Demonstration
 
     (otx) ...$ otx demo --help
-    usage: otx demo [-h] -i INPUT --load-weights LOAD_WEIGHTS [--fit-to-size FIT_TO_SIZE FIT_TO_SIZE] [--loop] [--delay DELAY] [--display-perf] [template] {params} ...
+    usage: otx demo [-h] -i INPUT --load-weights LOAD_WEIGHTS [--fit-to-size FIT_TO_SIZE FIT_TO_SIZE] [--loop] [--delay DELAY] [--display-perf] [--save-results-to SAVE_RESULTS_TO] [template] {params} ...
```
`save-results-to` is a good name, but what about `output`? Other CLI commands use the name `-o, --output`, so I think output is also a good idea.

GalyaZalesskaya replied:
Thank you, great point

GalyaZalesskaya replied:
Done

sovrasov commented on otx/cli/tools/demo.py:
```diff
@@ -100,12 +107,50 @@ def get_predictions(task, frame):
+        saved_frames[filename] = frame
+
+
+def dump_frames(saved_frames, output_path, capture):
```
Can we extract this function to some utils file (preferable is an existing one) to avoid duplication in exportable code?

GalyaZalesskaya replied:
Done

sovrasov reviewed (commented):
Could you also update README.md in the exportable code package according to comments from customer and the latest changes in CLI?

Commit:
Revert stream modification

Commit:
Merge branch 'develop' into gz/save_to_demo

Commit:
Revoke stream changes + change parameter name

Commit:
Remove unused imports

Commit:
Update exportable code demo

GalyaZalesskaya commented:
@eunwoosh @harimkang @sovrasov Thank you for your very useful comments. I removed extra output for the streamer and moved logic for saving images to one function in utils that CLI demo and exportable code demo share between each other.

@sovrasov I updated the exportable code README according to the latest CLI changes and customer feedback. The rest of the changes with exportable code from the feedback will be added in a separate PR to speed up the review of this one.

harimkang reviewed (dismissed):
LGTM,
BTW, Do you have any plan to add some test cases for new functions in demo/?

eunwoosh commented on otx/cli/tools/utils/demo/visualization.py:
```diff
@@ -161,3 +165,46 @@ def draw_predictions(task_type: TaskType, predictions: List[Annotation], frame:
+    if isinstance(input_path, int):
+        return []
+    if "DIR" in capture.get_type():
+        return [f.name for f in Path(input_path).iterdir() if f.is_file()]
```
It's for `DirStreamer` right? Then, I think this and that in `DirStreamer` should have same order.
I think one way to ensure that is a making a function to make a list of files in directory and make both of them use it.

GalyaZalesskaya replied:
I'm sorry, I didn't quite understand you.  Can you please elaborate?

eunwoosh replied:
As far as I understand, function `get_input_names_list` is in charge of providing list of names to save the frames.
And `if "DIR" in capture.get_type():` is for the case where `DirStreamer` is used. The problem is that each frame doesn't have information about which file is used for current frame in that case. So, when saving the frame, get filenames by `[f.name for f in Path(input_path).iterdir() if f.is_file()]` and then using each file name iterating `zip(filenames, saved_frames)`.
What I think as problem is an order of `[f.name for f in Path(input_path).iterdir() if f.is_file()]` in the function and order of `sorted(os.listdir(self.dir))` in `DirStreamer` can be different. So, I think both order of list should be same.
Of course, I may be misunderstanding. then I really appreciate if you let me know :)

Commit:
Fix pylint

GalyaZalesskaya commented:
> LGTM, BTW, Do you have any plan to add some test cases for new functions in demo/?

Sure, I'll add tests on Monday in this PR or in the separate PR (if we'll be able to merge it today)

goodsong81 reviewed (approved):
Thank you for the update. Could you please submit a PR with some tests?
